### PR TITLE
les: fix SendTx cost calculation and verify cost table

### PR DESCRIPTION
**Pull request by zsfelfoldi**
This PR backports the new transaction cost calculation (including the size cost correction for large transactions which is now used by latest servers too). It also adds a verification for required message types in the announced cost table.

### CODE REVIEW

File: les/txrelay.go [hunk 1]
```diff
@@ -113,21 +114,22 @@ func (self *LesTxRelay) send(txs types.Transactions, count int) {
...
 			canSend: func(dp distPeer) bool {
 				return dp.(*peer) == pp
 			},
 			request: func(dp distPeer) func() {
 				peer := dp.(*peer)
-				cost := peer.GetRequestCost(SendTxMsg, len(ll))
+				cost := peer.GetTxRelayCost(len(ll), len(enc))
 				peer.fcServer.QueueRequest(reqID, cost)
-				return func() { peer.SendTxs(reqID, cost, ll) }
+				return func() { peer.SendTxs(reqID, cost, enc) }
```
Comment (rjl493456442) [inline les/txrelay.go]: After modifying the `TestTransactionStatusLes2` to this style, the RLP decode is failed.

**Comment by zsfelfoldi** [comment]
I don't think we should strictly check the length of the cost table. We might want to add new requests later and in many cases it is doable without a new protocol version. Maybe we could add some protection against sending excessively large handshake packets in general but I don't think we should add it to 1.8.

**Comment by zsfelfoldi** [comment]
Even worse, currently the server does announce the cost of requests for all protocol versions. Now that you brought this up, this might not be ideal but expecting an exact size for the cost table would definitely break now.

**Comment by rjl493456442** [comment]
@zsfelfoldi You are right. We have to also consider the compatibility between the 1.8.26 light clients and old server.

File: les/peer.go [hunk 2]
```diff
@@ -163,7 +168,41 @@ func (p *peer) GetRequestCost(msgcode uint64, amount int) uint64 {
...
+	defer p.lock.RUnlock()
+
+	var msgcode uint64
+	switch p.version {
+	case lpv1:
+		msgcode = SendTxMsg
+	case lpv2:
+		msgcode = SendTxV2Msg
+	default:
+		panic(nil)
```
Comment (fjl) [inline les/peer.go]: Please add a message. `panic(nil)` is really bad style.

File: les/peer.go [hunk 3]
```diff
@@ -485,6 +524,20 @@ func (p *peer) Handshake(td *big.Int, head common.Hash, headNum uint64, genesis
...
 		p.fcServer = flowcontrol.NewServerNode(params)
 		p.fcCosts = MRC.decode()
+		var checkList []uint64
+		switch p.version {
+		case lpv1:
+			checkList = reqListV1
+		case lpv2:
+			checkList = reqListV2
+		default:
+			panic(nil)
```
Comment (fjl) [inline les/peer.go]: Same here.

File: les/peer.go (hunk 2 above)
Comment (zsfelfoldi) [reply les/peer.go]: these were all removed when LES/1 was removed so I really don't think we need to care about it in this hotfix

**Pull request merged**